### ISSUE TITLE
Improve documentation for custom cgroups (--cgroup-parent)

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#12849]
The `--cgroup-parent` option was proposed in https://github.com/docker/docker/issues/8551, and implemented in https://github.com/docker/docker/pull/11428 and (for `docker build`) in https://github.com/docker/docker/pull/11844

The documentation for this feature is currently very limited, and even *missing* in some places (for example, in the `docker run` [command line reference](https://github.com/docker/docker/blob/8bd5a95e1e39541dfc5dc635c5c8e7604fe10028/docs/sources/reference/commandline/cli.md#run)). In addition, there's no example on how to use this feature, which might help users understand on how to use this.

Obviously, care should be taken to not to explain cgroups in itself, but I found a nice short walk-through here; http://linuxaria.com/article/introduction-to-cgroups-the-linux-conrol-group)

**Comment by rncry** [comment]
+1 an example for this would be very useful

**Comment by hqhq** [comment]
@thaJeztah There are some related issues:
https://github.com/opencontainers/runc/issues/141
https://github.com/opencontainers/runc/pull/336
https://github.com/docker/docker/issues/16681
https://github.com/opencontainers/runc/pull/339

When we work them out, I can help improve the documentation. So #debs :)

**Comment by thaJeztah** [comment]
Thanks @hqhq, that's awesome!

**Comment by anusha-ragunathan** [comment]
@thaJeztah: This option is in-flux and currently being discussed for systemd cgroups. Meanwhile, here's something for cgroupfs.

cgroups subsystems are individually mounted under cgroup mount /sys/fs/cgroup. When cgroups are managed by cgroupfs, a cgroup is created per container for every subsystem, under the default parent cgroup, "docker". Typically hierarchy is /sys/fs/cgroup/<subsystem>/docker/containerID, where subsystem list is derived from /proc/cgroups.

The --cgroup-parent option can be used to override the default parent cgroup. A container that is run with a specific non-default parent is nested under the parent's directory in the cgroup fs hierarchy.

Start container c1.
`docker run -it busybox sh`
This creates container with ID, c1ID under default parent "docker" and a dir such as /sys/fs/cgroup/<subsystem>/docker/c1ID is created.

Start container c2 under the same cgroup as c1 above.
`docker run --cgroup-parent=c1ID busybox sh`
This creates c2's cgroups dir nested under c1 such as /sys/fs/cgroup/<subsystem>/docker/c1ID/c2ID

Another way to confirm that the child is nested under the parent, is to peek into the container's /proc/self/cgroup file.

**Comment by thaJeztah** [comment]
Thanks @anusha-ragunathan I'm good with waiting until things stabilize a bit; this may be a nice subject for a guest writer to write an article on, like we had for the storage driver section

/cc @moxiegirl (just thinking out loud here :smile:)

**Comment by moxiegirl** [comment]
@thaJeztah Guest writer or internal --- yes,  I agree that the `cgroups` material is just not where it needs to be. @anusha-ragunathan gives us a starting point (thank you helpful smart person) of the technical details.  I"d like to go further:

* why should I care about cgroups as a Docker user
* what kinds of things do cgroups impact and how can I adjust or control tune them
* if I had a problem in an application and that problem was related to cgroups what would I see
* how do I troubleshoot or fix the problem.
* are there alternatives or similar things to cgroups 

Some of that I'm sure is in those tickets....

**Comment by thaJeztah** [comment]
Yes, I think a section like that could be really helpful

**Comment by moxiegirl** [comment]
@aditirajagopal  Here is a very interesting challenge.  WDYT?

**Comment by aditirajagopal** [comment]
@moxiegirl thank you! this sounds like a good place to start :smile: -- will follow-up on IRC!

**Comment by aditirajagopal** [comment]
@moxiegirl been trying to catch you on IRC - when is a good time to chat about this?

**Comment by anusha-ragunathan** [comment]
https://github.com/opencontainers/runc/pull/336 and https://github.com/docker/docker/pull/18612 implement the "--cgroup-parent" option implementation on systemd. Feature is ready for documentation.

**Comment by moxiegirl** [comment]
@anusha-ragunathan I think I got that on #18379 wdyt?

**Comment by thaJeztah** [comment]
@moxiegirl just saw the text you wrote for https://github.com/docker/docker/pull/18379, 😍. After that is completed/merged, perhaps we should consider moving it to a central location so that we can link to it from the `docker daemon`, `docker run` and `docker create` docs? For example, to here; https://docs.docker.com/engine/reference/run/#specifying-custom-cgroups, that would reduce duplication. wdyt?

**Comment by hellobiek** [comment]
@anusha-ragunathan do you mean in readhat 6?
in my 7u test, the results is not what you had descripted.

**Comment by moxiegirl** [comment]
@thaJeztah I am more thinking we need a whole article just on cgroups --- rather than just adding to `run` or some other place.  Once we have a central article, we can link to the central article rather than attempting to coordinate the whole deal from various files.

**Comment by thaJeztah** [comment]
@moxiegirl oh, that would be great. I referred to the section in "run reference", because we're describing how to constraint resources in a container. cgroups are part of that (and, in fact, some of the other options are actually just wrappers for setting cgroups)

**Comment by moxiegirl** [comment]
@thaJeztah Yep, there is a whole thing around setting up the daemon and then overriding...it is really central and we are handling it poorly right now.  So much to do so little time. Some days...

![](http://mrwgifs.com/wp-content/uploads/2013/06/Penny-Needs-Help-For-Her-Computer-Addiction-On-The-Big-Bang-Theory.gif)

**Comment by thaJeztah** [comment]
![Moby Beacon](https://camo.githubusercontent.com/af70ca77855c8f568f51635edc5e827a15f222da/68747470733a2f2f692e696d67666c69702e636f6d2f6a6e716e312e6a7067)

**Comment by moxiegirl** [comment]
![image](https://cloud.githubusercontent.com/assets/1347209/11920361/aadfbfb0-a721-11e5-9b0a-641a421ffc9c.png)

**Comment by thaJeztah** [comment]
Awesome!! Haha!

**Comment by moxiegirl** [comment]
@aditirajagopal Are you still interested in doing this material?  The reason I ask is we have activity around this on another PR.  I know it is a big commitment so if you would like to take a pass and do something smaller, I can find you another.

What do you think?

**Comment by aditirajagopal** [comment]
@moxiegirl yeah I noticed. Maybe I can hold off working on this till it gets fleshed out more?

**Comment by drmalex07** [comment]
Hello Docker team,

This is also related to https://github.com/docker/docker/issues/16681, but since it's closed i report here.

### Description of problem:

I am trying to use `cgroup-parent` option while creating some containers in order to limit resource usage for a group of containers (not a single one). I provide an existing cgroup (nested properly into kernel's subsystems i am interested on), but the result is to get a newly created cgroup prefixed with the name i provided as input (for `cgroup-parent` option)

`docker version`:
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64


`docker info`:
Containers: 2
Images: 17
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 21
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 4
Total Memory: 494 MiB
Name: debian8-sandbox-1
ID: MONT:SGPY:7CEE:572J:TDLW:7E32:ARNI:ZVQL:4OR4:AKFI:J6EF:3OLV

`uname -a`:
Linux debian8-sandbox-1 3.16.0-4-amd64 #1 SMP Debian 3.16.7-ckt11-1+deb8u5 (2015-10-09) x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.): -

### Steps to Reproduce:


1. Create an arbitrary cgroup like `cluster-1`. E.g.:
```shell
> lscgroup | grep cluster-1
cpuset:/cluster-1
cpu,cpuacct:/cluster-1
memory:/cluster-1
devices:/cluster-1
net_cls,net_prio:/cluster-1
blkio:/cluster-1
```

2. Run a container inside `cluster-1`.
```shell
docker run --cgroup-parent cluster-1 -it --name redis-1 --hostname redis-1.internal redis:3.0
```

### Actual Results:

The cgroup i intended to use is ignored. Instead a new cgroup is created under `system.slice`. The truncated output of `systemd-cgls` is:
```shell
├─system.slice
....
│ ├─cluster-1-1b1a1b5d22a60a012ee785565b99830abb9b2b1c840abc233a5f42b9b69de578.scope
│ │ └─4422 redis-server *:637
....
``` 

### Expected Results:

I expected to group (>=1) containers into pre-existing cgroups. Maybe, i have misunderstood the semantics of `cgroup-parent` option, but i think that documentation is too terse at this point.  

### Additional info:

What is the proposed way of allocating resources for groups of containers? Should i reside on plain `cgrules.conf` functionality ?

**Comment by hqhq** [comment]
@drmalex07 Unfortunately `--cgroup-parent` can't have the save semantic for both "fs cgroup" and "systemd cgroup", you can see discussion here https://github.com/opencontainers/runc/pull/336

That's something we need in docs.

And I think the latest upstream docker will satisfy your scenario, because it'll use "fs cgroup" as the default cgroup manager.

**Comment by drmalex07** [comment]
Hello @hqhq, Ok, i undestand now. 

When you say "latest upstream" i you mean 1.9.1 or master?. My results above are tested on 1.9.1 and still docker doesn't seem to add (or reuse) a cgroupfs entry. Basically it always creates a scope (named as `docker-<cid>.scope`) under system.slice.

Let's assume i want to only manage cgroups by systemd means (i.e define slices and assign services to them). Is there a way to group running containers (wrapped as systemd service units) into a certain slice ? I think it would be sufficient to somehow support moving the `docker-<cid>.scope` (which should remain as a way to set per-container quotas) into a existing slice (e.g. `memory/cluster-1.slice/docker-<cid>.scope`)

**Comment by hqhq** [comment]
@drmalex07 "latest upstream" I mean master.

If you want to mange cgroup by systemd, you can start daemon with:
`docker daemon --exec-opt native.cgroupdriver=systemd --cgroup-parent system.slice`

Then all containers will have a default parent "system.slice", you can also change it when running a container with:
`docker run -ti --cgroup-parent my.slice ubuntu bash`

**Comment by drmalex07** [comment]
@hqhq 

That seems to work fine! The `1.10.0-dev` daemon (from master.dockerproject.org) correctly places container scopes inside given slices.

Do we know when version `1.10` will be released and when will be available via apt.dockerproject.org repositories?

**Comment by LK4D4** [comment]
@thaJeztah This should be done. And I think should be done by Docker Inc.

**Comment by michacassola** [comment]
Hey Guys, @moxiegirl @anusha-ragunathan @aditirajagopal @hqhq ,
is there any news about this years after? The documentation is still very short.

Can I use this:

> Start container c1.
> docker run -it busybox sh
> This creates container with ID, c1ID under default parent "docker" and a dir such as /sys/fs/cgroup//docker/c1ID is created.
> 
> Start container c2 under the same cgroup as c1 above.
> docker run --cgroup-parent=c1ID busybox sh
> This creates c2's cgroups dir nested under c1 such as /sys/fs/cgroup//docker/c1ID/c2ID

to set a general limits for all containers involved?
And can I use it in some way in a compose file?
